### DISCUSSION TITLE
manage memory consumption when supporting multiple language models

### DISCUSSION THREAD

**Discussion by lsmith77** [fastapi/fastapi#6052] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app = FastAPI()

model = {"en": spacy.load("en_core_web_sm"), "de": spacy.load("de_core_news_lg")}

@app.post("/check")
async def post_check_(text: str):
    lang = detect_language(text)
    tokens = model[lang](text)
    ..
```

### Description

I have a single endpoint that handles both English and German text using the spacy models. These models vary greatly in size and the German one specifically uses up 1GB+ of memory.

Having both models in the same worker, leads to a situation where many requests are being served for English (actually the ratio of English vs. German requests is 2:1), that are using all the memory of the German model.

As a quick solution I have build a solution to no longer load the German model and instead move to German model handling into a new HTTP endpoint. But I wonder if this is the right pattern. We do plan to add Spanish next year (and add more languages eventually).

We are offering a real time writing assistant (ie. we need to answer within a couple hundred milliseconds), so sometimes we get new requests before a previous one is answered. Also we have freemium and paid users. For these reasons I have pondered in the past if we should actually put all requests on a queue and the response is held until the message is processed on the queue. If we would do this, then we could have queue workers by language model. But I also worry that this might be over-engineering.

So are there any other patterns I should explore?

### Operating System

Linux

### FastAPI Version

0.88.0

### Python Version

3.11.0

**Comment by jgould22** [comment]
There are definitely options for queues etc.

Prior to that I might put the supported models behind a setting. You could then load models based on configuration.

Once you can choose at startup which models are loaded you can place a load balancer in front and send requests to English supporting processes for english requests, or requests for German to German supporting processes etc.

**Comment by iudeen** [comment]
A better scalable approach is how @jgould22 suggested, have each of your FastAPI application serve one language and a load balancer to route. 

You can use something like nginx for that and have your reverse proxy routes defined something like `/translate/en/app/...` where `en` is your identifier.

If that is not an option, you need to lazy load your models, this avoids having a higher consumption during the start, this however may increase when multiple users start using multiple languages.

**Comment by lsmith77** [comment, marked as answer]
> As a quick solution I have build a solution to no longer load the German model and instead move to German model handling into a new HTTP endpoint. But I wonder if this is the right pattern. We do plan to add Spanish next year (and add more languages eventually).

thank you @jgould22 and @iudeen. the above is essentially what you suggest. since most requests first need language detection and since the English model is so small, I basically now set it up like so:

1) “main app” (exposed to the outside)
* handles language detection
* handles authentication/authorization
* reads user/team configuration from redis
* applies the NLP model for English or calls “german app” if the request is for German
2) “german app” (not exposed to the outside)
* applies the NLP model for German

In principle I could break out English into its own app too but since it uses relatively small amounts of memory I will keep it with the “main app” to keep latency lower.

This pattern could be used to add new languages too.

It does complicate the local setup a bit, so for now in the local setup and on CI I will keep the single server setup and add some end2end tests to ensure that the multi server setup works.
